hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/utils/SparkSampleWritesUtils.java

```diff
@@ -0,0 +1,143 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi.client.utils;
+
+import org.apache.hudi.client.SparkRDDWriteClient;
+import org.apache.hudi.client.WriteStatus;
+import org.apache.hudi.client.common.HoodieSparkEngineContext;
+import org.apache.hudi.common.config.HoodieMetadataConfig;
+import org.apache.hudi.common.fs.FSUtils;
+import org.apache.hudi.common.model.HoodieCommitMetadata;
+import org.apache.hudi.common.model.HoodieRecord;
+import org.apache.hudi.common.model.HoodieTableType;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.hadoop.CachingPath;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.List;
+
+import static org.apache.hudi.common.table.HoodieTableMetaClient.SAMPLE_WRITES_FOLDER_PATH;
+import static org.apache.hudi.common.util.ValidationUtils.checkState;
+import static org.apache.hudi.config.HoodieCompactionConfig.COPY_ON_WRITE_RECORD_SIZE_ESTIMATE;
+import static org.apache.hudi.config.HoodieWriteConfig.SAMPLE_WRITES_ENABLED;
+import static org.apache.hudi.config.HoodieWriteConfig.SAMPLE_WRITES_SIZE;
+
+/**
+ * The utilities class is dedicated to estimating average record size by writing sample incoming records
+ * to `.hoodie/.aux/.sample_writes/<instant time>/<epoch millis>` and reading the commit metadata.
+ *
+ * TODO handle sample_writes sub-path clean-up w.r.t. rollback and insert overwrite. (HUDI-6044)
+ */
+public class SparkSampleWritesUtils {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SparkSampleWritesUtils.class);
+
+  public static void overwriteRecordSizeEstimateIfNeeded(JavaSparkContext jsc, JavaRDD<HoodieRecord> records, HoodieWriteConfig writeConfig, String instantTime) {
+    if (!writeConfig.getBoolean(SAMPLE_WRITES_ENABLED)) {
+      LOG.debug("Skip overwriting record size estimate as it's disabled.");
+      return;
+    }
+    HoodieTableMetaClient metaClient = getMetaClient(jsc, writeConfig.getBasePath());
+    if (metaClient.isTimelineNonEmpty()) {
+      LOG.info("Skip overwriting record size estimate due to timeline is non-empty.");
+      return;
+    }
+    try {
+      Pair<Boolean, String> result = doSampleWrites(jsc, records, writeConfig, instantTime);
+      if (result.getLeft()) {
+        long avgSize = getAvgSizeFromSampleWrites(jsc, result.getRight());
+        LOG.info("Overwriting record size estimate to " + avgSize);
+        writeConfig.setValue(COPY_ON_WRITE_RECORD_SIZE_ESTIMATE, String.valueOf(avgSize));
```

Comment:
generally our writeConfig is immutable. So, if we embed the call to do sample writes w/n UpsertPartitioner, we can avoid this.

Reply:
i don't see a big issue with overwriting this config at this point, as we're actually fixing a value which wasn't set properly. Not sure how exactly to embed it with upsertpartitioner, but i think this is a clean separation point to intercept and sample records. besides, people may have set a custom partitioner

Reply:
the PR is updated to keep writeconfig immutable